classes/module/Module.php

```diff
@@ -1045,6 +1060,125 @@ public function editExceptions($id_hook, $excepts)
         return $result;
     }
 
+    /**
+     * Install a tab according to its defined structure
+     *
+     * @param array $tab_info The structure of the tab.
+     *
+     * @return bool   true if the tab was installed successfully, false otherwise
+     *                (and a descriptive error will be added to $_errors)
+     */
+    private function installTab($tab_info)
+    {
+        $class_name = isset($tab_info['class_name']) ? $tab_info['class_name'] : null;
+        if (!$class_name) {
+            $this->_errors[] = Tools::displayError('Missing class name for tab "'.$tab_info['name'].'"');
+            return false;
+        }
+
+        $tab = new Tab();
+        $tab->module = $this->name;
+        $tab->class_name = $class_name;
+
+        // Default to active if it's not specified
+        $tab->active = (!isset($tab_info['active']) || $tab_info['active'] ? 1 : 0);
+
+        $tab->name = array();
+        foreach (Language::getLanguages(true) as $lang) {
+            $tab->name[$lang['id_lang']] = $tab_info['name'];
+        }
+
+        // Setup the parent relationship
+        if (isset($tab_info['parent_class'])) {
+            $tab->id_parent = Tab::getIdFromClassName($tab_info['parent_class']);
+            if (!$tab->id_parent) {
+                $this->_errors[] = Tools::displayError('Failed to find tab parent class "'.$tab_info['parent_class'].'"');
+                return false;
+            }
+        } elseif (isset($tab_info['hidden']) && $tab_info['hidden']) {
+            $tab->id_parent = -1;
+        } else {
+            $tab->id_parent = 0;
+        }
+
+        $success = $tab->add();
+        if (!$success) {
+            $this->_errors[] = Tools::displayError('Failed to install admin tab "'.$tab_info['name'].'"');
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Uninstalls a tab given its defined structure.
+     *
+     * @param array $tab_info The structure of the tab.
+     *
+     * @return bool Returns true if the specified tab was successfully deleted, false otherwise
+     *              (and a descriptive error will be added to $_errors).
+     */
+    private function uninstallTab($tab_info)
+    {
+        $class_name = isset($tab_info['class_name']) ? $tab_info['class_name'] : null;
+        if (!$class_name) {
+            $this->_errors[] = Tools::displayError('Missing class name for tab "'.$tab_info['name'].'"');
+            return false;
+        }
+
+        while ($id_tab = (int)Tab::getIdFromClassName($class_name)) {
```

Comment:
In case a tab was registered multiple times in error (e.g. when you're testing your module installation). This makes sure that, whatever happened at installation time, you don't leave broken admin tabs behind.